process/peer/process.go

```diff
@@ -345,6 +376,93 @@ func (p *validatorStatistics) getPeerAccount(address []byte) (state.PeerAccountH
 	return peerAccount, nil
 }
 
+// loadPreviousShardHeaders loads the previous shard headers for a given metablock. For the metachain it's easy
+//  since it has all the shard headers in its storage, but for the shard it's a bit trickier and we need
+//  to iterate through past metachain headers until we find all the ShardData we are interested in
+func (p *validatorStatistics) loadPreviousShardHeaders(currentHeader, previousHeader *block.MetaBlock) error {
+
+	p.mutPrevShardInfo.Lock()
+	defer p.mutPrevShardInfo.Unlock()
+
+	p.prevShardInfo = make(map[uint32]block.ShardData)
+	missingShardIds := make([]uint32, 0)
+
+	for _, currentShardData := range currentHeader.ShardInfo {
+		prevShardData := p.getMatchingShardData(currentShardData.ShardID, previousHeader.ShardInfo)
+		if prevShardData != nil {
+			p.prevShardInfo[currentShardData.ShardID] = *prevShardData
+		} else {
+			missingShardIds = append(missingShardIds, currentShardData.ShardID)
+		}
+	}
+
+	searchHeader := &block.MetaBlock{}
+	*searchHeader = *previousHeader
+	for len(missingShardIds) > 0 {
+		recursiveHeader, err := process.GetMetaHeader(searchHeader.GetPrevHash(), p.dataPool.MetaBlocks(), p.marshalizer, p.storageService)
+		if err != nil {
+			return err
+		}
+		for i, shardId := range missingShardIds {
+			prevShardData := p.getMatchingShardData(shardId, recursiveHeader.ShardInfo)
+			if prevShardData != nil {
+				p.prevShardInfo[shardId] = *prevShardData
+				missingShardIds = append(missingShardIds[:i], missingShardIds[i+1:]...)
+			}
+		}
+		*searchHeader = *recursiveHeader
+	}
+	return nil
+}
+
+func (p *validatorStatistics) loadPreviousShardHeadersMeta(header *block.MetaBlock) error {
+	p.mutPrevShardInfo.Lock()
+	defer p.mutPrevShardInfo.Unlock()
+
+	metaDataPool, ok := p.dataPool.(dataRetriever.MetaPoolsHolder)
+	if !ok {
+		return process.ErrInvalidMetaPoolHolder
+	}
+
+	for _, shardData := range header.ShardInfo {
+		if shardData.Nonce == 1 {
+			continue
+		}
+
+		previousHeader, err := process.GetShardHeader(shardData.PrevHash, metaDataPool.ShardHeaders(), p.marshalizer,
```

Comment:
can you reformat the parameter list? one per line if it exceeds the limit

Reply:
done